src/Functions/keyvaluepair/src/KeyValuePairExtractor.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <unordered_map>
+#include <string>
+
+namespace DB
+{
+/*
+ * Extracts key value pairs from strings. Strings do not need to be key-value pair only,
```

Comment:
Before finishing the code, we can discuss examples of how it would work from the user side.

Reply:
Initially, I overlooked `gtest_key_value_pair_extractor.cpp` and wanted to see any examples to understand what is `reserved_char` for example. But I've checked these examples, and they look sane.

For now, the question is, what are the escaping rules for quoted string? E.g., are the escaping of control chars applied ( `"\n"` ) ? And how are quotes handled (with slash `"with \"quoted\" inside"`, or csv-style double quotes `"with ""quotes"" inside"`)?

Also, are there any plans to support the automatic detection of separators? This function tries to do best-effort parsing and be flexible, so it may work like that. So all `parserKV('aa: one,  bb: two,'), parserKV('aa=one bb=two'), parserKV('aa => one; bb => two')` would return `{"aa: "one", "bb": "two"}`. For that, we can have some list of possible kv_separators (`':', '=', '=>', ...`) and item delimiters (`',', ' ', ';', '|', ...`) and check what combination is appeared more frequently. But still trying to figure out if it would be helpful or if it would be challenging to implement and use then.

Reply:
1. **For now, the question is, what are the escaping rules for quoted string?** - _I am not sure I understand the question, but escaping does not happen on quoted strings. It'll read everything in until it gets another unescaped quote._
2. **Are the escaping of control chars applied ( "\n" ) ?** - _It'll read everything in, which means the output will contain the back-slash._
3.  **And how are quotes handled (with slash "with \"quoted\" inside", or csv-style double quotes "with ""quotes"" inside")?** - _Quotes, just like any other special character, need to be escaped. The escaping character is configurable, but it defaults to back-slash (\\)._
4. **Also, are there any plans to support the automatic detection of separators?** - _It is not on my plans and I think we should keep the first version simpler. I can totally implement it in the future or as a follow up of this PR if we decide it's actually useful._

I still haven't had the time to look into parameter cross validation (e.g user setting back-slash as both key_value_delimiter and escape character).

Reply:
> It'll read everything in until it gets another unescaped quote

So, having quote `"`" in quoted text is not possible?

Reply:
It is possible, but it needs to be escaped. Assuming the escaping character is '\',  an input like: "random_key"="aaa\\"bbb" will output: "random_key"="aaa"bbb". The quote in the middle being a value character, not a delimiter